### PR TITLE
fix(DTFS2-6888): deal validation before submitting to ACBS

### PR DESCRIPTION
**Pull request by abhi-markan**
## Introduction :pencil2:
Ensure deals are validated prior to submission to ACBS.

## Resolution :heavy_check_mark:
* Introduced a singular validation function `canSubmitToACBS` which performs myriads of check on the most up-to-date TFM deal before triggering a request to `ACBS`. Functions name have been updated to correctly describe respective function responsiblities and returned response.
* Refactored code.
* Improved test coverage.

## Miscellaneous :heavy_plus_sign:
* Added missing `--fix` flag to `lint:fix` scripts alongside with quotes.
* Replaced jest deprecated function `toBeCalledTimes` with `toHaveBeenCalledTimes`.
* Dependencies updates.
* Added temporary IDs to constants.
* Improved documentation.

### CODE REVIEW

File: trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+const submitDealToACBS = async (deal, firstSubmissionCheck = true, validateOnly = false) => {
+  try {
+    const { _id, dealSnapshot, tfm } = deal;
+    const acceptable = [CONSTANTS.DEALS.SUBMISSION_TYPE.AIN, CONSTANTS.DEALS.SUBMISSION_TYPE.MIN];
+
+    console.info('⚡ Validating ACBS checks for deal %s', _id);
+
+    const validSubmissionType = acceptable.includes(dealSnapshot?.submissionType);
+    const validFirstSubmission = firstSubmissionCheck ? !tfm?.acbs : true;
+    const validIds = await dealHasAllValidUkefIds(_id);
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js:25]: `dealHasAllValidUkefIds` has incorrect docs btw causing my vscode to get an error when you `await` it :) 

should be 

```
 * @returns {Promise(boolean)}  Boolean value
```
instead of
```
 * @returns {Boolean}  Boolean value
```

I think :)

File: trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * This function performs following checks:
+ * 1. Valid submission type either AIN or MIN
+ * 2. Valid first submission check, ensuring `tfm.acbs` property does not exist
+ * 3. Valid deal and facility IDs
+ * @param {Object} deal - TFM deal object
+ * @param {boolean} [firstSubmissionCheck=true] - Flag indicating whether to perform the first submission check, defaults to `true`
+ * @param {boolean} [validateOnly=false] - Flag indicating whether to just to perform validation, defaults to `false`
+ * @returns {boolean} - True if the deal can be submitted to ACBS, otherwise false.
+ */
+const submitDealToACBS = async (deal, firstSubmissionCheck = true, validateOnly = false) => {
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js:16]: I really like the validation on this function.

Would it be best to split this into two different functions? Someone in a few years time might find it surprising to see `submitDealToAcbs(bla,true, true)` would not actually submit a deal to acbs I think?

What do you think about something like the below

```
const getIsValidDealToSubmit = async (deal, isFirstSubmission) => {
  console.info('⚡ Validating ACBS checks for deal %s', _id);

  const { _id, dealSnapshot, tfm } = deal;

  const acceptableSubmissionTypes = [CONSTANTS.DEALS.SUBMISSION_TYPE.AIN, CONSTANTS.DEALS.SUBMISSION_TYPE.MIN];
  const isValidSubmissionType = acceptableSubmissionTypes.includes(dealSnapshot?.submissionType);
  const isValidFirstSubmission = isFirstSubmission ? !tfm?.acbs : true;
  const isValidIds = await dealHasAllValidUkefIds(_id);
  const isValidDealToSubmit = isValidSubmissionType && isValidFirstSubmission && isValidIds;

  return isValidDealToSubmit;
};

const submitDealToACBS = async ({ deal, isFirstSubmission = true }) => {
  try {
    const { _id } = deal;
    const isValidDealToSubmit = getIsValidDealToSubmit(deal, isFirstSubmission);
    if (!isValidDealToSubmit) {
      throw new InvalidDealError({ dealId: _id }); // these could just be standard Errors
    }
    await submitACBSIfAllPartiesHaveUrn(_id);
  } catch (error) {
    console.error('Unable to submit deal to ACBS %o', error); // We shouldn't really need this -- it should log fine by just throwing the errors
    throw new UnableToSubmitDealToAcbsError({ dealId: _id, cause: error }); // these could just be standard Errors
  }
};
```

File: trade-finance-manager-api/src/v1/controllers/party.controller.js [hunk 3]
```diff
@@ -1,31 +1,33 @@
...
+    const tfmDeal = await api.updateDeal({ dealId, dealUpdate, auditDetails });
+
+    console.log('====5', tfmDeal);
 
-    if (updatedDeal.dealSnapshot) {
-      if (canDealBeSubmittedToACBS(updatedDeal.dealSnapshot.submissionType)) {
-        await submitACBSIfAllPartiesHaveUrn(dealId);
-      }
+    if (tfmDeal) {
+      await submitDealToACBS(tfmDeal);
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/controllers/party.controller.js:21]: not your code again but return of `submitDealToACBS` should be a promise

File: trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js (hunk 2 above)
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js:16]: Just a side note -- if we are defaulting `isFirstSubmission` to true, could we remove it entirely and add in a check where we call it when needed ie:

```
if(isFirstSubmission){
submitDealToACBS(deal)
}
```

Im honestly not very bothered either way, just thinking it might be nicer to do this check outside of the function unless we're doing it in multiple places

File: trade-finance-manager-api/src/v1/controllers/deal.submit.controller.js [hunk 4]
```diff
@@ -197,6 +185,13 @@ const submitDealAfterUkefIds = async (dealId, dealType, checker, auditDetails) =
       console.info('TFM deal %s stage has been updated to %s', dealId, updatedDealStage);
     }
 
+    console.log('===2', mappedDeal);
+    await submitDealToACBS(mappedDeal);
```
[the remark this answers is not in the document]
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/controllers/deal.submit.controller.js:189]: non review comment - i feel this funciton adds so so much added cognative load when you have to touch it (it's huge!)... Next time we need to address tech debt I feel this would be a great place to start

File: trade-finance-manager-api/src/v1/controllers/acbs.controller.js [hunk 5]
```diff
@@ -150,25 +155,32 @@ const checkAzureAcbsFunction = async () => {
...
-   * !isIssued(facilityStageInAcbs) && !facility.tfm.acbs.issuedFacilityMaster
-   * const facilityStageInAcbs = facility.tfm.acbs && facility.tfm.acbs.facilityStage;
+   * ACBS verification: Facility stage verification ensures facility which are
+   * under commitment stage (06) are only eligible for issuance
    */
-
+  console.error('Issuing facilities in ACBS for deal %s', deal._id);
   const acbsIssuedFacilitiesPromises = deal.facilities
-    .filter((facility) => facility.hasBeenIssued)
+    .filter((facility) => facility?.hasBeenIssued && !isIssuedInACBS(facility?.tfm?.acbs.facilityStage))
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/controllers/acbs.controller.js:183]: Sorry -- This is my lack of understanding with this as a general business concept -- if the facility has been issued but the tfm object has not been issued in acbs then it's issued?

**Review by AlexBramhill — COMMENTED** [review]
looking good

File: trade-finance-manager-api/src/v1/controllers/party.controller.js (hunk 3 above)
Comment (abhi-markan) [reply trade-finance-manager-api/src/v1/controllers/party.controller.js:27]: Have updated.

File: trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js (hunk 1 above)
Comment (abhi-markan) [reply trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js:29]: Have updated.

File: trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js (hunk 2 above)
Comment (abhi-markan) [reply trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js:16]: There were two separate functions before, however I have added these checks on the submission to avoid a bug or an eventuality where submission to ACBS can be invoked without any pre-requesite checks.

Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js:16]: I think we can still have the verification called inside the submit deal function, but keeping it seperate means we can call it seperately too unless I'm misunderstanding something?

ie:

```
const validateDealCanBeSubmittedToAcbs = async (deal, isFirstSubmission) => {
  // Validation stuff
    if (!isValidDealToSubmit) {
      throw new InvalidDealError({ dealId: _id });
    }
};

const submitDealToACBS = async ({ deal, isFirstSubmission = true }) => {
  try {
    validateDealCanBeSubmittedToAcbs(deal, isFirstSubmission); // we call the validation check here
    // Other logic
  } catch (error) {
    throw new UnableToSubmitDealToAcbsError({ dealId: _id, cause: error });
  }
};
```



------------------------------------



One of the reasons I'm slightly worried is I find this in `deal.submit.controller.js` line 184 very hard to understand without understanding that `submitDealToACBS` does not actually `submitDealToAcbs` the second time it's called, which is surprising. The function is both being used for `submitDealToACBS` and `validateDealCanBeSubmittedToAcbs` which I feel are two distinct functions (`submitDealToAcbs` can call `VAlidateDealCanBeSubmittedToAcbs`, as shown above)...

`deal.submit.controller.js` line 184:
```
    await submitDealToACBS(tfmDeal);

    if (await submitDealToACBS(tfmDeal, false, true)) {
      await issueAcbsFacilities(dealUpdate);
    }

    return tfmDeal;
```

Instead we could have:

```
try {
    await submitDealToAcbs(mappedDeal)
} catch(error) {
    if (error.cause instanceof InvalidDealError) {
        await issueAcbsFacilities(uploadDeal)
    } else {
        throw error
    }
}
```

We could even look to wrap that in it's own helper/pull it out into a service level function so 
```
    await submitDealToACBS(tfmDeal);

    if (await submitDealToACBS(tfmDeal, false, true)) {
      await issueAcbsFacilities(dealUpdate);
    }

    return tfmDeal;
```
becomes
```
await submitDealOrIssueDealFacilities(tfmDeal, dealUpdate)
```
and in a helper/service we have
```
try {
    await submitDealToAcbs(mappedDeal)
} catch(error) {
    if (error.cause instanceof InvalidDealError) {
        await issueAcbsFacilities(uploadDeal)
    } else {
        throw error
    }
}
```

tldr I'm a little worried about having a funciton saying `doSomething` and if we have `doSomething(true)` it actually does `somethingElse` and doesn't `doSomething`

**Review by AlexBramhill — COMMENTED** [review]
Sorry for the incredibly long comment, happy to have a chat about it. agree with everything in theory -- just a little worried about function names

File: trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js (hunk 2 above)
Comment (abhi-markan) [reply trade-finance-manager-api/src/v1/helpers/submit-deal-acbs.js]: Have refactoerd. Thanks

File: trade-finance-manager-api/src/v1/controllers/deal.submit.controller.js [hunk 6]
```diff
@@ -82,56 +83,53 @@ const submitDealAfterUkefIds = async (dealId, dealType, checker, auditDetails) =
...
-       */
-      return api.updateDeal({ dealId, dealUpdate: updatedDealWithTasks, auditDetails });
+    // Update the deal with all the above modifications
+    const tfmDeal = api.updateDeal({ dealId, dealUpdate, auditDetails });
+
+    // Submit to ACBS
+    const canSubmitDealToACBS = await canSubmitToACBS(tfmDeal);
+
+    if (canSubmitDealToACBS) {
+      await createACBS(tfmDeal);
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/controllers/deal.submit.controller.js:119]: love this

File: trade-finance-manager-api/src/v1/controllers/acbs.controller.js [hunk 7]
```diff
@@ -8,46 +8,69 @@ const tfmController = require('./tfm.controller');
...
 
   return false;
 };
 
+/**
+ * Creates an ACBS task and adds it to the ACBS log.
+ * @param {object} deal - The deal object containing the dealSnapshot property with the bank details.
+ * @returns {Promise<Boolean>} - True if the ACBS task is successfully created and added to the log, false otherwise.
+ */
 const createACBS = async (deal) => {
```
Comment (AlexBramhill) [inline trade-finance-manager-api/src/v1/controllers/acbs.controller.js:56]: do we want to name this createACBSTast?

**Review by avaitonis — APPROVED** [review]
LGTM

File: trade-finance-manager-api/api-tests/v1/deals/deals-submit.second-submission.api-test.js [hunk 8]
```diff
@@ -31,6 +24,17 @@ const MOCK_GEF_DEAL_MIN = require('../../../src/v1/__mocks__/mock-gef-deal-MIN')
 const { submitDeal, createSubmitBody } = require('../utils/submitDeal');
 const { mockFindOneDeal, mockUpdateDeal } = require('../../../src/v1/__mocks__/common-api-mocks');
 
+jest.mock('../../../src/v1/controllers/acbs.controller', () => ({
+  issueAcbsFacilities: jest.fn(),
+  createACBS: jest.fn(),
+}));
+
+jest.mock('../../../src/v1/controllers/deal.controller', () => ({
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply trade-finance-manager-api/api-tests/v1/deals/deals-submit.second-submission.api-test.js:32]: Yes correct, have removed.

File: trade-finance-manager-api/src/v1/helpers/can-submit-to-acbs.test.js [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+    expect(consoleInfoMock).toHaveBeenCalledTimes(2);
+    expect(consoleInfoMock).toHaveBeenCalledWith('⚡ Validating ACBS deal %s prerequisites.', mockTfmDeal._id);
+    expect(consoleInfoMock).toHaveBeenCalledWith('✅ Deal %s is eligible for submission to ACBS.', mockTfmDeal._id);
+
+    expect(result).toEqual(true);
+  });
+
+  const invalidDeals = [null, undefined, {}, { ukefDealId: '123' }, [], 123, '123', '!$%'];
+
+  it.each(invalidDeals)('should throw an error if the deal object is invalid', async (deal) => {
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply trade-finance-manager-api/src/v1/helpers/can-submit-to-acbs.test.js:83]: Good spot, have rectified.

**Review by ChristianMcc — APPROVED** [review]
